File: src/gromacs/mdrun/integrator.cpp
/*
 * This file is part of the GROMACS molecular simulation package.
 *
 * Copyright (c) 2018, by the GROMACS development team, led by
 * Mark Abraham, David van der Spoel, Berk Hess, and Erik Lindahl,
 * and including many others, as listed in the AUTHORS file in the
 * top-level source directory and at http://www.gromacs.org.
 *
 * GROMACS is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public License
 * as published by the Free Software Foundation; either version 2.1
 * of the License, or (at your option) any later version.
 *
 * GROMACS is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with GROMACS; if not, see
 * http://www.gnu.org/licenses, or write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA.
 *
 * If you want to redistribute modifications to GROMACS, please
 * consider that scientific software is very special. Version
 * control is crucial - bugs must be traceable. We will be happy to
 * consider code for inclusion in the official distribution, but
 * derived work must not be called official GROMACS. Details are found
 * in the README & COPYING files - if they are missing, get the
 * official version at http://www.gromacs.org.
 *
 * To help us fund GROMACS development, we humbly ask that you cite
 * the research papers on the package. Check out http://www.gromacs.org.
 */
/*! \internal
 * \brief Defines the dispatch function for the .mdp integrator field.
 *
 * \author David van der Spoel <david.vanderspoel@icm.uu.se>
 * \author Mark Abraham <mark.j.abraham@gmail.com>
 * \ingroup module_mdrun
 */
#include "gmxpre.h"

#include "integrator.h"

#include <cassert>

#include "gromacs/compat/make_unique.h"
#include "gromacs/mdtypes/md_enums.h"
#include "gromacs/utility/exceptions.h"

#include "md.h"

namespace gmx
{

//! \brief Run the correct integrator function.
void IntegratorDispatcher::run()
{
    unsigned int ei {
        getMethod()
    };
    switch (ei)
    {
        case eiSteep:
            do_steep();
            break;
        case eiCG:
            do_cg();
            break;
        case eiNM:
            do_nm();
            break;
        case eiLBFGS:
            do_lbfgs();
            break;
        case eiTPI:
        case eiTPIC:
            if (!EI_TPI(ei))
            {
                GMX_THROW(APIError("do_tpi integrator would be called for a non-TPI integrator"));
            }
            do_tpi();
            break;
        case eiSD2_REMOVED:
            GMX_THROW(NotImplementedError("SD2 integrator has been removed"));
        default:
            GMX_THROW(APIError("Non existing integrator selected"));
    }
}

IIntegrator::~IIntegrator() = default;

SimulationMethod IntegratorDispatcher::getMethod() const
{
    return method_;
}

IntegratorBuilder::DataSentry
IntegratorBuilder::Base::setAggregateAdapter(std::unique_ptr<IntegratorAggregateAdapter> adapter)
{
    (void)adapter;
    // default no-op. There should probably be some sanity checking here.
    return {};
}

IntegratorBuilder::Base::~Base() = default;

class IntegratorDispatcherBuilder : public IntegratorBuilder::Base
{
    public:
        /*!
         * \brief To avoid ambiguity, this type can only be instantiated with
         * its main parameter initialized.
         *
         * \{
         */
        IntegratorDispatcherBuilder() = delete;
        IntegratorDispatcherBuilder(const IntegratorDispatcherBuilder &)                = delete;
        IntegratorDispatcherBuilder &operator=(const IntegratorDispatcherBuilder &)     = delete;
        // Move cannot be implicitly handled. We would need to copy the const method_ member.
        IntegratorDispatcherBuilder(IntegratorDispatcherBuilder &&) noexcept            = delete;
        IntegratorDispatcherBuilder &operator=(IntegratorDispatcherBuilder &&) noexcept = delete;
        /* \} */

        /*!
         * \brief Initialize a new builder.
         *
         * \param method Simulation method that our dispatcher will dispatch to.
         */
        explicit IntegratorDispatcherBuilder(SimulationMethod method) :
            method_ {method}
        {
            // For some reason at least one linter thinks this constructor is never used,
            // but it is used by IntegratorBuilder::create
        }

        /*!
         * \brief Create the dispatcher with the catch-all parameter pack.
         */
        IntegratorBuilder::DataSentry setAggregateAdapter(std::unique_ptr<IntegratorAggregateAdapter> container)
        override
        {
            if (paramsContainer_)
            {
                GMX_THROW(APIError("setParams has already been called on this builder."));
            }
            else
            {
                paramsContainer_ = std::move(container);
            }
            IntegratorBuilder::DataSentry sentry;
            // This is probably where we should bind the integrator and data sentry when we have an API to do so.
            return sentry;
        }

        std::unique_ptr<IIntegrator> build() override
        {
            // Indicates a usage error, not a violation of an established invariant.
            if (!paramsContainer_)
            {
                GMX_THROW(APIError("This builder requires setParams() to be called before build()."));
            }

            auto integrator = gmx::compat::make_unique<IntegratorDispatcher>(method_, *paramsContainer_);
            paramsContainer_ = nullptr;
            return integrator;
        }


    private:
        /*!
         * \brief Method that our dispatcher will be dispatching for.
         */
        const SimulationMethod method_;
        /*!
         * \brief Dispatcher instance is not created until setParams is called.
         */
        std::unique_ptr<IntegratorAggregateAdapter> paramsContainer_ {nullptr};
};

IntegratorBuilder::~IntegratorBuilder() = default;

IntegratorBuilder::IntegratorBuilder(IntegratorBuilder &&) noexcept = default;

IntegratorBuilder &IntegratorBuilder::operator=(IntegratorBuilder &&) noexcept = default;

std::unique_ptr<IIntegrator> IntegratorBuilder::build()
{
    return impl_->build();
}

IntegratorBuilder::IntegratorBuilder(std::unique_ptr<::gmx::IntegratorBuilder::Base> impl) :
    impl_ {std::move(impl)}
{
    if (!impl_)
    {
        GMX_THROW(APIError("Builder should not be created with an uninitialized implementation object."));
    }
}


IntegratorBuilder IntegratorBuilder::create(const SimulationMethod &integratorType)
{
    // Dispatch creation of an appropriate builder for the integration method.
    // Initially, the factory is not extensible and only has one catch-all builder
    // to dispatch to.
    std::unique_ptr<IntegratorBuilder::Base> builderImpl;
    auto method = static_cast<decltype(eiMD)>(integratorType.method_);
    switch (method)
    {
        case eiMD:
        case eiBD:
        case eiSD1:
        case eiVV:
        case eiVVAK:
            if (!EI_DYNAMICS(method))
            {
                GMX_THROW(APIError("do_md integrator would be called for a non-dynamical integrator"));
            }
            builderImpl = gmx::compat::make_unique<gmx::MDIntegrator::Builder>();
            break;
        default:
            builderImpl = gmx::compat::make_unique<IntegratorDispatcherBuilder>(integratorType);
    }

    IntegratorBuilder builder {
        std::move(builderImpl)
    };

    assert(builder.impl_);
    return builder;
}

IntegratorBuilder::DataSentry::~DataSentry() = default;
IntegratorBuilder::DataSentry::DataSentry()  = default;
IntegratorBuilder::DataSentry::DataSentry(IntegratorBuilder::DataSentry &&) noexcept = default;
IntegratorBuilder::DataSentry &
IntegratorBuilder::DataSentry::operator=(IntegratorBuilder::DataSentry &&) noexcept = default;

}  // namespace gmx

File: tests/physicalvalidation/physical_validation/util/integrator.py
###########################################################################
#                                                                         #
#    physical_validation,                                                 #
#    a python package to test the physical validity of MD results         #
#                                                                         #
#    Written by Michael R. Shirts <michael.shirts@colorado.edu>           #
#               Pascal T. Merz <pascal.merz@colorado.edu>                 #
#                                                                         #
#    Copyright (C) 2012 University of Virginia                            #
#              (C) 2017 University of Colorado Boulder                    #
#                                                                         #
#    This library is free software; you can redistribute it and/or        #
#    modify it under the terms of the GNU Lesser General Public           #
#    License as published by the Free Software Foundation; either         #
#    version 2.1 of the License, or (at your option) any later version.   #
#                                                                         #
#    This library is distributed in the hope that it will be useful,      #
#    but WITHOUT ANY WARRANTY; without even the implied warranty of       #
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU    #
#    Lesser General Public License for more details.                      #
#                                                                         #
#    You should have received a copy of the GNU Lesser General Public     #
#    License along with this library; if not, write to the                #
#    Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,     #
#    Boston, MA 02110-1301 USA                                            #
#                                                                         #
###########################################################################
r"""
This module contains low-level functionality of the
`physical_validation.integrator` module. The functions in this module should
generally not be called directly. Please use the high-level functions from
`physical_validation.integrator`.
"""
import numpy as np

from . import plot


def calculate_rmsd(data, time=None, slope=False):
    assert isinstance(data, np.ndarray) and data.ndim == 1
    assert time is None or isinstance(time, np.ndarray) and time.ndim == 1

    avg = data.mean()

    if time is None:
        time = np.arange(data.size)

    fit = np.polyfit(time, data, 1)

    def f(x): return fit[0]*x + fit[1]

    if slope:
        rmsd = 0
        for t, d in zip(time, data):
            rmsd += (d - f(t))**2
        rmsd = np.sqrt(rmsd / data.size)
    else:
        rmsd = data.std()

    return avg, rmsd, fit[0]


def max_deviation(dts, rmsds):
    dt_ratio_2 = (dts[:-1] / dts[1:])**2
    rmsds = rmsds[:-1] / rmsds[1:]
    return np.max(np.abs(1 - rmsds/dt_ratio_2))


def check_convergence(const_traj,
                      convergence_test=max_deviation,
                      verbose=True, slope=False,
                      screen=False, filename=None):

    assert isinstance(const_traj, dict)
    assert len(const_traj) >= 2

    if verbose:
        print('{:65s}'.format('-'*65))
        print('{:>10s} {:>10s} {:>10s} {:>10s} {:^21s}'.format('dt', 'avg', 'rmsd', 'slope', 'ratio'))
        print('{:43s} {:>10s} {:>10s}'.format('', 'dt^2', 'rmsd'))
        print('{:65s}'.format('-'*65))

    prev = None

    results = {}
    for dt, traj in sorted(const_traj.items(), key=lambda x: float(x[0]), reverse=True):
        assert isinstance(traj, np.ndarray)
        assert traj.ndim == 1 or traj.ndim == 2
        dt = float(dt)

        if traj.ndim == 1:
            data = traj
            time = None
        else:
            data = traj[1]
            time = traj[0]

        results[dt] = calculate_rmsd(data, time, slope)

        if verbose:
            if prev is None:
                print('{:10.4g} {:10.2f} {:10.2e} {:10.2e} {:>10s} {:>10s}'.format(dt, results[dt][0], results[dt][1],
                                                                                   results[dt][2], '--', '--'))
                prev = [dt, results[dt][1]]
            else:
                print('{:10.4g} {:10.2f} {:10.2e} {:10.2e} {:10.2f} {:10.2f}'.format(dt, results[dt][0], results[dt][1],
                                                                                     results[dt][2],
                                                                                     prev[0]**2/dt**2,
                                                                                     prev[1]/results[dt][1]))
                prev = [dt, results[dt][1]]

    if verbose:
        print('{:65s}'.format('-'*65))

    dts = np.sort(np.array([float(dt) for dt in results.keys()]))[::-1]
    rmsds = np.array([float(results[dt][1]) for dt in dts])

    do_plot = screen or filename is not None

    if do_plot:
        data = [{'x': dts[1:],
                 'y': rmsds[:-1] / rmsds[1:],
                 'name': 'Integrator convergence'},
                {'x': dts[1:],
                 'y': (dts[:-1] / dts[1:])**2,
                 'name': 'Expected convergence'}]

        plot.plot(data,
                  legend='best',
                  title='Actual vs. expected convergence',
                  xlabel='Time step',
                  ylabel='Convergence',
                  xlim=(0, dts[1]),
                  inv_x=True,
                  filename=filename,
                  screen=screen)

    return convergence_test(dts, rmsds)
